// @openshift/dynamic-plugin-sdk-webpack/dist/index.cjs.js
/**
  OpenShift Dynamic Plugin SDK
  https://github.com/openshift/dynamic-plugin-sdk

  packageName    : @openshift/dynamic-plugin-sdk-webpack
  packageVersion : 5.1.1
  buildDate      : April 16, 2026
  buildTime      : 5:43:42 PM GMT
  gitCommit      : e590f3155a460913f5e8c42ec1c8f900fd1588d7
  gitBranch      : main
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var yup = require('yup');
var semver = require('semver');
var identity = require('lodash/identity');
var isEmpty = require('lodash/isEmpty');
var mapValues = require('lodash/mapValues');
var intersection = require('lodash/intersection');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var identity__default = /*#__PURE__*/_interopDefaultLegacy(identity);
var isEmpty__default = /*#__PURE__*/_interopDefaultLegacy(isEmpty);
var mapValues__default = /*#__PURE__*/_interopDefaultLegacy(mapValues);
var intersection__default = /*#__PURE__*/_interopDefaultLegacy(intersection);

const DEFAULT_REMOTE_ENTRY_CALLBACK = '__load_plugin_entry__';

// TODO(vojtech): suppress false positive https://github.com/jsx-eslint/eslint-plugin-react/pull/3326
/**
 * Schema for a valid semver string.
 */
const semverStringSchema = yup.string()
    .required()
    .test('semver-string', 'Must be a strictly valid semver string', (value) => {
    // valid may return a cleaned version (e.g., by stripping out leading 'v') but we need value to always be clean
    return semver.valid(value, { loose: false }) === value;
});
/**
 * Schema for a valid plugin name.
 *
 * @example
 * ```
 * foo
 * foo-bar
 * foo.bar
 * foo.bar-Test
 * Foo-Bar-abc.123
 * ```
 */
const pluginNameSchema = yup.string()
    .required()
    .matches(/^[a-zA-Z]+(?:[-.]?[a-zA-Z0-9]+)*$/);
/**
 * Schema for a valid extension type.
 *
 * @example
 * ```
 * app.foo
 * app.foo-bar
 * app.foo/bar
 * My-app.Foo-Bar
 * My-app.Foo-Bar/abcTest
 * ```
 */
const extensionTypeSchema = yup.string()
    .required()
    .matches(/^[a-zA-Z]+(?:-[a-zA-Z]+)*\.[a-zA-Z]+(?:-[a-zA-Z]+)*(?:\/[a-zA-Z]+(?:-[a-zA-Z]+)*)*$/);
/**
 * Schema for a valid feature flag name.
 *
 * @example
 * ```
 * FOO
 * FOO_BAR
 * FOO_BAR123
 * ```
 */
const featureFlagNameSchema = yup.string()
    .required()
    .matches(/^[A-Z]+[A-Z0-9_]*$/);
/**
 * Schema for `Extension` objects.
 */
const extensionSchema = yup.object()
    .required()
    .shape({
    type: extensionTypeSchema,
    properties: yup.object().required(),
    flags: yup.object().shape({
        required: yup.array().of(featureFlagNameSchema),
        disallowed: yup.array().of(featureFlagNameSchema),
    }),
});
/**
 * Schema for an array of `Extension` objects.
 */
const extensionArraySchema = yup.array().of(extensionSchema).required();
/**
 * Schema for `Record<string, string>` objects.
 */
const recordStringStringSchema = yup.object() // Rejects non-objects and null
    .test('Record<string, string> | undefined', 'Must be an object with string keys and string values OR undefined', (obj) => {
    // Allow undefined because these fields are optional
    if (obj === undefined) {
        return true;
    }
    // Objects can have Symbol() as keys, so ensure there are none
    if (Object.getOwnPropertySymbols(obj).length > 0) {
        return false;
    }
    // Object keys can only be symbols or strings, but since we've ruled out symbols,
    // we can assume that all keys are strings. We just need to check the values now.
    return Object.values(obj).every((value) => typeof value === 'string');
});
/**
 * Schema for `Record<string, string>` objects where the values are valid semver ranges.
 */
const recordStringSemverRangeSchema = recordStringStringSchema.test('Record<string, semver_range> | undefined', 'Must be an object with string keys and semver range string values OR undefined', (obj) => {
    // Allow undefined because these fields are optional
    if (obj === undefined) {
        return true;
    }
    // recordStringStringSchema ensures that all keys and values are strings,
    // so we just need to check that all values are valid semver ranges now.
    return Object.values(obj).every((value) => semver.validRange(value) !== null);
});
/**
 * Schema for `PluginRuntimeMetadata` objects.
 */
const pluginRuntimeMetadataSchema = yup.object().required().shape({
    name: pluginNameSchema,
    version: semverStringSchema,
    dependencies: recordStringSemverRangeSchema,
    optionalDependencies: recordStringSemverRangeSchema,
    customProperties: yup.object(),
});
/**
 * Schema for `RemotePluginManifest` objects.
 */
pluginRuntimeMetadataSchema.shape({
    baseURL: yup.string().required(),
    extensions: extensionArraySchema,
    loadScripts: yup.array().of(yup.string().required()).required(),
    registrationMethod: yup.string().oneOf(['callback', 'custom']).required(),
    buildHash: yup.string(),
});

/**
 * Schema for `PluginBuildMetadata` objects.
 */
const pluginBuildMetadataSchema = pluginRuntimeMetadataSchema.shape({
    exposedModules: recordStringStringSchema,
});
/**
 * Schema for `PluginModuleFederationSettings` objects.
 */
const pluginModuleFederationSettingsSchema = yup.object().required().shape({
    libraryType: yup.string(),
    sharedScopeName: yup.string(),
});
/**
 * Schema for `PluginEntryCallbackSettings` objects.
 */
const pluginEntryCallbackSettingsSchema = yup.object().required().shape({
    name: yup.string(),
    pluginID: yup.string(),
});
/**
 * Schema for adapted `DynamicRemotePluginOptions` objects.
 */
const dynamicRemotePluginAdaptedOptionsSchema = yup.object().required().shape({
    pluginMetadata: pluginBuildMetadataSchema,
    extensions: extensionArraySchema,
    sharedModules: yup.object().required(),
    moduleFederationSettings: pluginModuleFederationSettingsSchema,
    entryCallbackSettings: pluginEntryCallbackSettingsSchema,
    entryScriptFilename: yup.string().required(),
    pluginManifestFilename: yup.string().required(),
});

const findPluginChunks = (containerName, compilation) => {
    const allChunks = Array.from(compilation.chunks);
    const entryChunk = allChunks.find((chunk) => chunk.name === containerName);
    if (!entryChunk) {
        throw new Error(`Cannot find entry chunk ${containerName}`);
    }
    if (entryChunk.hasRuntime()) {
        return { entryChunk };
    }
    const runtimeChunk = allChunks.find((chunk) => {
        const chunkName = chunk.name ?? '';
        // Chunk runtime can be undefined | string | SortableSet<string>
        // so we need a different check for the SortableSet value
        if (typeof entryChunk.runtime === 'string') {
            return entryChunk.runtime === chunkName;
        }
        else if (entryChunk.runtime) {
            return entryChunk.runtime.has(chunkName);
        }
    });
    if (!runtimeChunk) {
        throw new Error(`Cannot find runtime chunk for entry chunk ${containerName}`);
    }
    return { entryChunk, runtimeChunk };
};
const getChunkFiles = (chunk, compilation, includeFile = (assetInfo) => !assetInfo.development && !assetInfo.hotModuleReplacement) => Array.from(chunk.files).filter((fileName) => {
    const asset = compilation.getAsset(fileName);
    if (!asset) {
        throw new Error(`Missing asset data for ${fileName}`);
    }
    return includeFile(asset.info);
});

class GenerateManifestPlugin {
    constructor(options) {
        this.options = options;
    }
    apply(compiler) {
        const { containerName, manifestFilename, manifestData, transformManifest } = this.options;
        const { publicPath } = compiler.options.output;
        if (!publicPath) {
            throw new Error('output.publicPath option must be set to ensure plugin assets are loaded properly in the browser');
        }
        compiler.hooks.thisCompilation.tap(GenerateManifestPlugin.name, (compilation) => {
            compilation.hooks.processAssets.tap({
                name: GenerateManifestPlugin.name,
                // Using one of the later asset processing stages to ensure all assets
                // are already added and optimized within the given webpack compilation
                stage: compiler.webpack.Compilation.PROCESS_ASSETS_STAGE_ANALYSE,
            }, () => {
                const { entryChunk, runtimeChunk } = findPluginChunks(containerName, compilation);
                const pluginChunks = runtimeChunk ? [runtimeChunk, entryChunk] : [entryChunk];
                const loadScripts = pluginChunks.reduce((acc, chunk) => [...acc, ...getChunkFiles(chunk, compilation)], []);
                const manifest = transformManifest({
                    ...manifestData,
                    baseURL: compilation.getAssetPath(publicPath, {}),
                    loadScripts,
                    buildHash: compilation.fullHash,
                });
                const manifestContent = JSON.stringify(manifest, null, compiler.options.mode === 'production' ? undefined : 2);
                compilation.emitAsset(manifestFilename, new compiler.webpack.sources.RawSource(Buffer.from(manifestContent)));
                const warnings = [];
                if (manifest.extensions.length === 0) {
                    warnings.push('Plugin has no extensions');
                }
                if (!manifest.baseURL.endsWith('/')) {
                    warnings.push('Plugin base URL (output.publicPath) should have a trailing slash');
                }
                warnings.forEach((message) => {
                    const error = new compiler.webpack.WebpackError(message);
                    error.file = manifestFilename;
                    compilation.warnings.push(error);
                });
            });
        });
    }
}

class PatchEntryCallbackPlugin {
    constructor(options) {
        this.options = options;
    }
    apply(compiler) {
        const { containerName, callbackName, pluginID } = this.options;
        compiler.hooks.thisCompilation.tap(PatchEntryCallbackPlugin.name, (compilation) => {
            compilation.hooks.processAssets.tap({
                name: PatchEntryCallbackPlugin.name,
                stage: compiler.webpack.Compilation.PROCESS_ASSETS_STAGE_OPTIMIZE,
            }, () => {
                const { entryChunk } = findPluginChunks(containerName, compilation);
                getChunkFiles(entryChunk, compilation).forEach((fileName) => {
                    compilation.updateAsset(fileName, (source) => {
                        const newSource = new compiler.webpack.sources.ReplaceSource(source);
                        const fromIndex = source.source().toString().indexOf(`${callbackName}(`);
                        if (fromIndex >= 0) {
                            newSource.insert(fromIndex + callbackName.length + 1, `'${pluginID}', `);
                        }
                        else {
                            const error = new compiler.webpack.WebpackError(`Missing call to ${callbackName}`);
                            error.file = fileName;
                            error.chunk = entryChunk;
                            compilation.errors.push(error);
                        }
                        return newSource;
                    });
                });
            });
        });
    }
}

class ValidateCompilationPlugin {
    constructor(options) {
        this.options = options;
    }
    apply(compiler) {
        const { containerName, jsonpLibraryType } = this.options;
        compiler.hooks.done.tap(ValidateCompilationPlugin.name, ({ compilation }) => {
            const { runtimeChunk } = findPluginChunks(containerName, compilation);
            if (runtimeChunk) {
                const errorMessage = jsonpLibraryType
                    ? 'Detected separate runtime chunk while using jsonp library type.\n' +
                        'This configuration is not allowed since it will cause issues when reloading plugins at runtime.\n' +
                        'Please update your webpack configuration to avoid emitting a separate runtime chunk.'
                    : 'Detected separate runtime chunk while using non-jsonp library type.\n' +
                        'This configuration is not recommended since it may cause issues when reloading plugins at runtime.\n' +
                        'Consider updating your webpack configuration to avoid emitting a separate runtime chunk.';
                const error = new compiler.webpack.WebpackError(errorMessage);
                error.chunk = runtimeChunk;
                (jsonpLibraryType ? compilation.errors : compilation.warnings).push(error);
            }
        });
    }
}

const DEFAULT_MANIFEST = 'plugin-manifest.json';
const DEFAULT_ENTRY_SCRIPT = 'plugin-entry.js';
class DynamicRemotePlugin {
    constructor(options) {
        this.adaptedOptions = {
            pluginMetadata: options.pluginMetadata,
            extensions: options.extensions,
            sharedModules: options.sharedModules ?? {},
            moduleFederationSettings: options.moduleFederationSettings ?? {},
            entryCallbackSettings: options.entryCallbackSettings ?? {},
            entryScriptFilename: options.entryScriptFilename ?? DEFAULT_ENTRY_SCRIPT,
            pluginManifestFilename: options.pluginManifestFilename ?? DEFAULT_MANIFEST,
            transformPluginManifest: options.transformPluginManifest ?? identity__default["default"],
        };
        try {
            dynamicRemotePluginAdaptedOptionsSchema.validateSync(this.adaptedOptions, {
                strict: true,
                abortEarly: false,
            });
        }
        catch (e) {
            throw new Error(`Invalid ${DynamicRemotePlugin.name} options:\n` +
                e.errors.join('\n'));
        }
        const overlapDependencyNames = intersection__default["default"](Object.keys(this.adaptedOptions.pluginMetadata.optionalDependencies ?? {}), Object.keys(this.adaptedOptions.pluginMetadata.dependencies ?? {}));
        if (overlapDependencyNames.length > 0) {
            throw new Error(`Detected overlap between dependencies and optionalDependencies: ${overlapDependencyNames.join(', ')}`);
        }
    }
    apply(compiler) {
        const { pluginMetadata, extensions, sharedModules, moduleFederationSettings, entryCallbackSettings, entryScriptFilename, pluginManifestFilename, transformPluginManifest, } = this.adaptedOptions;
        const containerName = pluginMetadata.name;
        const moduleFederationLibraryType = moduleFederationSettings.libraryType ?? 'jsonp';
        const moduleFederationSharedScope = moduleFederationSettings.sharedScopeName ?? 'default';
        const { ModuleFederationPlugin = compiler.webpack.container.ModuleFederationPlugin, ContainerPlugin = compiler.webpack.container.ContainerPlugin, } = moduleFederationSettings.pluginOverride ?? {};
        const entryCallbackName = entryCallbackSettings.name ?? DEFAULT_REMOTE_ENTRY_CALLBACK;
        const entryCallbackPluginID = entryCallbackSettings.pluginID ?? pluginMetadata.name;
        const jsonp = moduleFederationLibraryType === 'jsonp';
        const containerLibrary = {
            type: moduleFederationLibraryType,
            name: jsonp ? entryCallbackName : containerName,
        };
        const containerModules = mapValues__default["default"](pluginMetadata.exposedModules ?? {}, (moduleRequest, moduleName) => ({
            import: moduleRequest,
            name: `exposed-${moduleName}`,
        }));
        // Assign a unique name for the webpack build
        compiler.options.output.uniqueName ??= containerName;
        // Generate webpack federated module container assets
        new ModuleFederationPlugin({
            name: containerName,
            library: containerLibrary,
            filename: entryScriptFilename,
            exposes: containerModules,
            shared: sharedModules,
            shareScope: moduleFederationSharedScope,
        }).apply(compiler);
        // ModuleFederationPlugin does not generate a container entry when the provided
        // exposes option is empty; we fix that by invoking the ContainerPlugin manually
        if (isEmpty__default["default"](containerModules)) {
            new ContainerPlugin({
                name: containerName,
                library: containerLibrary,
                filename: entryScriptFilename,
                exposes: containerModules,
                shareScope: moduleFederationSharedScope,
            }).apply(compiler);
        }
        // Generate plugin manifest
        new GenerateManifestPlugin({
            containerName,
            manifestFilename: pluginManifestFilename,
            manifestData: {
                name: pluginMetadata.name,
                version: pluginMetadata.version,
                dependencies: pluginMetadata.dependencies,
                optionalDependencies: pluginMetadata.optionalDependencies,
                customProperties: pluginMetadata.customProperties,
                extensions,
                registrationMethod: jsonp ? 'callback' : 'custom',
            },
            transformManifest: transformPluginManifest,
        }).apply(compiler);
        // Post-process container entry generated by ModuleFederationPlugin
        if (jsonp) {
            new PatchEntryCallbackPlugin({
                containerName,
                callbackName: entryCallbackName,
                pluginID: entryCallbackPluginID,
            }).apply(compiler);
        }
        // Validate webpack compilation
        new ValidateCompilationPlugin({ containerName, jsonpLibraryType: jsonp }).apply(compiler);
    }
}

exports.DynamicRemotePlugin = DynamicRemotePlugin;
