// @openshift-console/dynamic-plugin-sdk-webpack/lib/coderefs/coderef-resolver.js
"use strict";
/* eslint-disable no-console */
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveExtension = exports.parseEncodedCodeRefValue = exports.setCodeRefError = exports.getCodeRefError = exports.isCodeRefError = exports.isExecutableCodeRef = exports.isEncodedCodeRef = void 0;
const dynamic_plugin_sdk_1 = require("@openshift/dynamic-plugin-sdk");
const lodash_1 = require("lodash");
const object_1 = require("../utils/object");
const promise_1 = require("../utils/promise");
/**
 * Extract the SDK's internal CodeRef symbol by applying it to a dummy function.
 *
 * This ensures we can detect code refs created by the SDK, which uses its own
 * private Symbol instance.
 */
const codeRefSymbol = Object.getOwnPropertySymbols((0, dynamic_plugin_sdk_1.applyCodeRefSymbol)(() => Promise.resolve()))[0];
if (!codeRefSymbol) {
    throw new Error('Failed to extract CodeRef symbol from the SDK');
}
const isEncodedCodeRef = (obj) => (0, lodash_1.isPlainObject)(obj) &&
    (0, lodash_1.isEqual)(Object.getOwnPropertyNames(obj), ['$codeRef']) &&
    typeof obj.$codeRef === 'string';
exports.isEncodedCodeRef = isEncodedCodeRef;
const isExecutableCodeRef = (obj) => (0, lodash_1.isFunction)(obj) &&
    (0, lodash_1.isEqual)(Object.getOwnPropertySymbols(obj), [codeRefSymbol]) &&
    obj[codeRefSymbol] === true;
exports.isExecutableCodeRef = isExecutableCodeRef;
const codeRefErrorSymbol = Symbol('error');
const isCodeRefError = (ref) => !!ref[codeRefErrorSymbol];
exports.isCodeRefError = isCodeRefError;
const getCodeRefError = (ref) => ref[codeRefErrorSymbol];
exports.getCodeRefError = getCodeRefError;
const setCodeRefError = (ref, e) => {
    ref[codeRefErrorSymbol] = e;
    return ref;
};
exports.setCodeRefError = setCodeRefError;
/**
 * Parse the `EncodedCodeRef` value into `[moduleName, exportName]` tuple.
 *
 * Returns an empty array if the value doesn't match the expected format.
 */
const parseEncodedCodeRefValue = (value) => {
    const match = value.match(/^([^.]+)(?:\.(.+)){0,1}$/);
    return match ? [match[1], match[2] || 'default'] : [];
};
exports.parseEncodedCodeRefValue = parseEncodedCodeRefValue;
/**
 * Returns an extension with its `CodeRef` properties replaced with referenced objects.
 *
 * The resulting Promise resolves with a new extension instance; its `properties` object
 * is cloned in order to preserve the original extension.
 */
const resolveExtension = async (extension) => {
    const clonedProperties = (0, lodash_1.cloneDeep)(extension.properties);
    const valueResolutions = [];
    (0, object_1.deepForOwn)(clonedProperties, exports.isExecutableCodeRef, (ref, key, obj) => {
        if ((0, exports.isCodeRefError)(ref)) {
            throw (0, exports.getCodeRefError)(ref);
        }
        valueResolutions.push(ref()
            .then((resolvedValue) => {
            obj[key] = resolvedValue;
            if ((0, lodash_1.isNil)(resolvedValue)) {
                console.warn(`Code reference property '${key}' resolved to null or undefined`);
            }
        })
            .catch((e) => {
            (0, exports.setCodeRefError)(ref, e ?? true);
            return e;
        }));
    });
    await (0, promise_1.settleAllPromises)(valueResolutions);
    // Return a new extension object with the resolved properties
    return {
        ...extension,
        properties: clonedProperties,
    };
};
exports.resolveExtension = resolveExtension;
