// @openshift-console/dynamic-plugin-sdk-webpack/lib/webpack/ConsoleRemotePlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsoleRemotePlugin = exports.dynamicModuleImportTransformFilter = exports.dynamicModulePackageSpecs = exports.validateConsoleExtensionsFileSchema = void 0;
const path = require("path");
const dynamic_plugin_sdk_webpack_1 = require("@openshift/dynamic-plugin-sdk-webpack");
const glob = require("glob");
const _ = require("lodash");
const readPkg = require("read-pkg");
const semver = require("semver");
const constants_1 = require("../constants");
const shared_modules_meta_1 = require("../shared-modules/shared-modules-meta");
const jsonc_1 = require("../utils/jsonc");
const schema_1 = require("../utils/schema");
const ExtensionValidator_1 = require("../validation/ExtensionValidator");
const SchemaValidator_1 = require("../validation/SchemaValidator");
const ValidationResult_1 = require("../validation/ValidationResult");
const DynamicModuleImportPlugin_1 = require("./DynamicModuleImportPlugin");
const loadPluginPackageJSON = () => readPkg.sync({ normalize: false });
const loadVendorPackageJSON = (moduleName) => 
// eslint-disable-next-line @typescript-eslint/no-var-requires
require(`${moduleName}/package.json`);
const getVendorPackageVersion = (moduleName) => {
    try {
        return loadVendorPackageJSON(moduleName).version;
    }
    catch (e) {
        return undefined;
    }
};
const getPackageDependencies = (pkg) => ({
    ...pkg.devDependencies,
    ...pkg.dependencies,
});
const getPluginSDKPackagePeerDependencies = () => loadVendorPackageJSON('@openshift-console/dynamic-plugin-sdk').peerDependencies;
const getPatternFlyStyles = (baseDir) => glob.sync(`${baseDir}/node_modules/@patternfly/react-styles/**/*.css`);
/**
 * Get webpack shared module configuration to use by Console plugins.
 *
 * This includes Console provided {@link sharedPluginModules} and shared dynamic modules
 * resolved from {@link dynamicModulePackageSpecs}.
 *
 * @see https://webpack.js.org/plugins/module-federation-plugin/#sharing-hints
 */
const getWebpackSharedModules = (pkg, dynamicModuleMaps) => {
    const sdkPkgDeps = getPluginSDKPackagePeerDependencies();
    const consoleProvidedSharedModules = shared_modules_meta_1.sharedPluginModules.reduce((acc, moduleName) => {
        const { singleton, allowFallback } = (0, shared_modules_meta_1.getSharedModuleMetadata)(moduleName);
        const providedVersionRange = sdkPkgDeps[moduleName];
        const moduleConfig = { singleton };
        if (!allowFallback) {
            moduleConfig.import = false;
        }
        if (semver.validRange(providedVersionRange)) {
            moduleConfig.requiredVersion = providedVersionRange;
        }
        acc[moduleName] = moduleConfig;
        return acc;
    }, {});
    const sharedDynamicModules = Object.entries(dynamicModuleMaps).reduce((acc, [moduleName, dynamicModuleMap]) => {
        const pluginDeps = getPackageDependencies(pkg);
        const moduleVersion = getVendorPackageVersion(moduleName);
        const moduleVersionRange = pluginDeps[moduleName];
        const moduleConfig = {};
        if (semver.valid(moduleVersion)) {
            moduleConfig.version = moduleVersion;
        }
        if (semver.validRange(moduleVersionRange)) {
            moduleConfig.requiredVersion = moduleVersionRange;
        }
        Object.values(dynamicModuleMap).forEach((request) => {
            acc[`${moduleName}/${request}`] = moduleConfig;
        });
        return acc;
    }, {});
    return { ...consoleProvidedSharedModules, ...sharedDynamicModules };
};
/**
 * Perform (additional) build-time validation of Console plugin metadata.
 *
 * Note that `DynamicRemotePlugin` takes care of basic build metadata validation.
 * Therefore, this function only performs additional Console specific validations.
 */
const validateConsoleBuildMetadata = (metadata) => {
    const result = new ValidationResult_1.ValidationResult('Console plugin metadata');
    result.assertions.validDNSSubdomainName(metadata.name, 'metadata.name');
    return result;
};
const validateConsoleExtensionsFileSchema = (extensions, description = 'console-extensions.json') => {
    const schema = (0, schema_1.loadSchema)('console-extensions.json');
    return new SchemaValidator_1.SchemaValidator(description).validate(schema, extensions);
};
exports.validateConsoleExtensionsFileSchema = validateConsoleExtensionsFileSchema;
const getDeprecatedSharedModuleWarnings = (pkg) => {
    const pluginDeps = getPackageDependencies(pkg);
    const warnings = [];
    shared_modules_meta_1.sharedPluginModules.forEach((moduleName) => {
        const { deprecated } = (0, shared_modules_meta_1.getSharedModuleMetadata)(moduleName);
        if (deprecated && pluginDeps[moduleName]) {
            warnings.push(`[DEPRECATION WARNING] Console provided shared module ${moduleName} has been deprecated: ${deprecated}`);
        }
    });
    return warnings;
};
const validateConsoleProvidedSharedModules = (pkg) => {
    const pluginDeps = getPackageDependencies(pkg);
    const sdkPkgDeps = getPluginSDKPackagePeerDependencies();
    const result = new ValidationResult_1.ValidationResult('package.json');
    shared_modules_meta_1.sharedPluginModules.forEach((moduleName) => {
        const { allowFallback } = (0, shared_modules_meta_1.getSharedModuleMetadata)(moduleName);
        // Skip modules that allow a fallback version to be provided by the plugin.
        // Also skip modules which are not explicitly listed in the plugin's dependencies.
        if (allowFallback || !pluginDeps[moduleName]) {
            return;
        }
        const providedVersionRange = sdkPkgDeps[moduleName];
        const consumedVersion = getVendorPackageVersion(moduleName);
        if (semver.validRange(providedVersionRange) && semver.valid(consumedVersion)) {
            result.assertThat(semver.satisfies(consumedVersion, providedVersionRange), `Console provides shared module ${moduleName} ${providedVersionRange} but plugin uses version ${consumedVersion}`);
        }
    });
    return result;
};
/**
 * PatternFly packages that support dynamic modules to be used with webpack module federation.
 *
 * Console provided {@link sharedPluginModules} should _NOT_ be listed here.
 */
const dynamicModulePatternFlyPackages = [
    '@patternfly/react-core',
    '@patternfly/react-data-view',
    '@patternfly/react-icons',
    '@patternfly/react-table',
    '@patternfly/react-templates',
];
/**
 * Default shared dynamic module package definitions.
 */
exports.dynamicModulePackageSpecs = dynamicModulePatternFlyPackages.reduce((acc, moduleName) => ({ ...acc, [moduleName]: {} }), {});
const dynamicModuleImportTransformFilter = (moduleRequest) => {
    const isCode = /\.(jsx?|tsx?)$/.test(moduleRequest);
    const isVendor = moduleRequest.includes('/node_modules/');
    return isCode && (!isVendor || moduleRequest.includes('/node_modules/@openshift-console/'));
};
exports.dynamicModuleImportTransformFilter = dynamicModuleImportTransformFilter;
/**
 * Generates Console dynamic plugin remote container and related assets.
 *
 * Refer to `console-dynamic-plugin-sdk/src/shared-modules.ts` for details on Console provided
 * shared modules and their configuration.
 *
 * @see {@link sharedPluginModules}
 * @see {@link getSharedModuleMetadata}
 */
class ConsoleRemotePlugin {
    constructor(options = {}) {
        this.baseDir = process.cwd();
        this.pkg = loadPluginPackageJSON();
        this.adaptedOptions = {
            pluginMetadata: options.pluginMetadata ?? this.pkg.consolePlugin,
            extensions: options.extensions ?? (0, jsonc_1.parseJSONC)(path.resolve(this.baseDir, constants_1.extensionsFile)),
            validateExtensionSchema: options.validateExtensionSchema ?? true,
            validateExtensionIntegrity: options.validateExtensionIntegrity ?? true,
            validateSharedModules: options.validateSharedModules ?? true,
            sharedDynamicModuleSettings: options.sharedDynamicModuleSettings ?? {},
        };
        if (this.adaptedOptions.validateExtensionSchema) {
            (0, exports.validateConsoleExtensionsFileSchema)(this.adaptedOptions.extensions).report();
        }
        if (this.adaptedOptions.validateSharedModules) {
            validateConsoleProvidedSharedModules(this.pkg).report();
        }
        validateConsoleBuildMetadata(this.adaptedOptions.pluginMetadata).report();
        const overlapDependencyNames = _.intersection(Object.keys(this.adaptedOptions.pluginMetadata.dependencies ?? {}), Object.keys(this.adaptedOptions.pluginMetadata.optionalDependencies ?? {}));
        if (overlapDependencyNames.length > 0) {
            throw new Error(`Detected overlap between dependencies and optionalDependencies: ${overlapDependencyNames.join(', ')}`);
        }
        const resolvedModulePaths = this.adaptedOptions.sharedDynamicModuleSettings.modulePaths ?? [
            path.resolve(process.cwd(), 'node_modules'),
        ];
        this.dynamicModuleMaps = (0, DynamicModuleImportPlugin_1.resolveDynamicModuleMaps)(this.adaptedOptions.sharedDynamicModuleSettings.packageSpecs ?? exports.dynamicModulePackageSpecs, resolvedModulePaths);
    }
    apply(compiler) {
        const { pluginMetadata, extensions, validateExtensionIntegrity, sharedDynamicModuleSettings, } = this.adaptedOptions;
        const { name, version, dependencies, optionalDependencies, customProperties, exposedModules, displayName, description, disableStaticPlugins, } = pluginMetadata;
        const logger = compiler.getInfrastructureLogger(ConsoleRemotePlugin.name);
        // Dynamic plugin assets should be loaded from /api/plugins/<plugin-name> endpoint.
        // Console Bridge server will fetch the asset from the appropriate plugin web server.
        const publicPath = `/api/plugins/${name}/`;
        if (compiler.options.output.publicPath !== undefined) {
            logger.warn(`output.publicPath is defined, but will be overridden to ${publicPath}`);
        }
        compiler.options.output.publicPath = publicPath;
        compiler.options.resolve = compiler.options.resolve ?? {};
        compiler.options.resolve.alias = compiler.options.resolve.alias ?? {};
        // Prevent PatternFly styles from being included in the compilation
        getPatternFlyStyles(this.baseDir).forEach((cssFile) => {
            if (Array.isArray(compiler.options.resolve.alias)) {
                compiler.options.resolve.alias.push({ name: cssFile, alias: false });
            }
            else {
                compiler.options.resolve.alias[cssFile] = false;
            }
        });
        const sharedModules = getWebpackSharedModules(this.pkg, this.dynamicModuleMaps);
        new dynamic_plugin_sdk_webpack_1.DynamicRemotePlugin({
            pluginMetadata: {
                name,
                version,
                dependencies,
                optionalDependencies,
                customProperties: _.merge({}, customProperties, {
                    console: { displayName, description, disableStaticPlugins },
                }),
                exposedModules,
            },
            extensions,
            sharedModules,
            entryCallbackSettings: {
                name: constants_1.REMOTE_ENTRY_CALLBACK,
            },
            entryScriptFilename: process.env.NODE_ENV === 'production'
                ? 'plugin-entry.[fullhash].min.js'
                : 'plugin-entry.js',
        }).apply(compiler);
        new DynamicModuleImportPlugin_1.DynamicModuleImportPlugin({
            dynamicModuleMaps: this.dynamicModuleMaps,
            moduleFilter: sharedDynamicModuleSettings.moduleFilter ?? exports.dynamicModuleImportTransformFilter,
        }).apply(compiler);
        // Post-build validations performed before emitting assets
        compiler.hooks.emit.tap(ConsoleRemotePlugin.name, (compilation) => {
            if (validateExtensionIntegrity) {
                const result = new ExtensionValidator_1.ExtensionValidator('Console plugin extensions').validate(compilation, extensions, exposedModules ?? {}, path.dirname(path.resolve(this.baseDir, constants_1.extensionsFile)));
                if (result.hasErrors()) {
                    const error = new compiler.webpack.WebpackError('ExtensionValidator has reported errors');
                    error.details = result.formatErrors();
                    error.file = constants_1.extensionsFile;
                    compilation.errors.push(error);
                }
            }
            getDeprecatedSharedModuleWarnings(this.pkg).forEach((message) => {
                compilation.warnings.push(new compiler.webpack.WebpackError(message));
            });
        });
    }
}
exports.ConsoleRemotePlugin = ConsoleRemotePlugin;
