// @openshift-console/dynamic-plugin-sdk-webpack/lib/webpack/DynamicModuleImportPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DynamicModuleImportPlugin = exports.resolveDynamicModuleMaps = void 0;
const fs = require("fs");
const path = require("path");
const chalk_1 = require("chalk");
const _ = require("lodash");
const dynamic_module_parser_1 = require("../utils/dynamic-module-parser");
const isDynamicModuleMap = (obj) => _.isPlainObject(obj) && Object.values(obj).every((value) => typeof value === 'string');
/**
 * Parse or generate dynamic module maps for the provided packages.
 */
const resolveDynamicModuleMaps = (
/** Dynamic module package definitions. */
packageSpecs, 
/** Absolute paths to `node_modules` directories to search. */
modulePaths) => Object.entries(packageSpecs).reduce((acc, [pkgName, { dynamicModuleDir = 'dist/dynamic', dynamicModuleMap = 'dist/dynamic-modules.json', indexModule = 'dist/esm/index.js', resolutionField = 'module', },]) => {
    const basePath = modulePaths
        .map((p) => path.resolve(p, pkgName))
        .find((p) => fs.existsSync(p) && fs.statSync(p).isDirectory());
    if (!basePath) {
        // eslint-disable-next-line no-console
        console.warn(`Cannot resolve base path for package ${pkgName}`);
        return acc;
    }
    const dynamicModuleMapPath = path.resolve(basePath, dynamicModuleMap);
    if (fs.existsSync(dynamicModuleMapPath)) {
        try {
            const obj = JSON.parse(fs.readFileSync(dynamicModuleMapPath, 'utf-8'));
            if (!isDynamicModuleMap(obj)) {
                throw new Error('Invalid dynamic module map object');
            }
            // eslint-disable-next-line no-console
            console.info(`${chalk_1.default.bold(pkgName)} dynamic modules taken from ${chalk_1.default.green(dynamicModuleMap)}`);
            acc[pkgName] = obj;
            return acc;
        }
        catch (e) {
            // eslint-disable-next-line no-console
            console.warn(`Failed to parse ${chalk_1.default.bold(pkgName)} dynamic modules`, e);
        }
    }
    const obj = (0, dynamic_module_parser_1.getDynamicModuleMap)(basePath, dynamicModuleDir, indexModule, resolutionField);
    // eslint-disable-next-line no-console
    console.info(`${chalk_1.default.bold(pkgName)} dynamic modules generated from ${chalk_1.default.green(dynamicModuleDir)}`);
    acc[pkgName] = obj;
    return acc;
}, {});
exports.resolveDynamicModuleMaps = resolveDynamicModuleMaps;
/**
 * Transforms import statements using Console dynamic module import loader.
 *
 * The loader is referenced using a module specifier since webpack takes care of creating
 * and initializing all loaders.
 */
class DynamicModuleImportPlugin {
    constructor(options) {
        this.options = options;
    }
    apply(compiler) {
        const { loader = '@openshift-console/dynamic-plugin-sdk-webpack/lib/webpack/loaders/dynamic-module-import-loader', dynamicModuleMaps, moduleFilter, } = this.options;
        compiler.hooks.thisCompilation.tap(DynamicModuleImportPlugin.name, (compilation) => {
            const modifiedModules = [];
            compiler.webpack.NormalModule.getCompilationHooks(compilation).beforeLoaders.tap(DynamicModuleImportPlugin.name, (_loaders, normalModule) => {
                const { userRequest } = normalModule;
                const moduleRequest = userRequest.substring(userRequest.lastIndexOf('!') === -1 ? 0 : userRequest.lastIndexOf('!') + 1);
                if (!modifiedModules.includes(moduleRequest) && moduleFilter(moduleRequest)) {
                    const loaderOptions = {
                        dynamicModuleMaps,
                        resourceMetadata: { jsx: /\.(jsx|tsx)$/.test(moduleRequest) },
                        skipImportPrefixes: [
                            // Imports for PatternFly deprecated APIs
                            '@patternfly/react-core/deprecated',
                            // Imports for PatternFly internal APIs (not exposed via package index)
                            '@patternfly/react-icons/dist/esm/createIcon',
                            '@patternfly/react-core/dist/esm/components/Tooltip/',
                            '@patternfly/react-core/dist/esm/components/Popover/',
                        ],
                    };
                    normalModule.loaders.push({ loader, options: loaderOptions });
                    modifiedModules.push(moduleRequest);
                }
            });
        });
    }
}
exports.DynamicModuleImportPlugin = DynamicModuleImportPlugin;
